xds/internal/xdsclient/transport/lrs/lrs_stream.go

```diff
@@ -0,0 +1,326 @@
+}
+
+// NewStream creates a new LRS Stream with the provided options.
+//
+// The actual streaming RPC call is initiated when the first call to ReportLoad
+// is made, and is terminated when the last call to ReportLoad is canceled.
+func NewStream(opts StreamOpts) *Stream {
+	lrs := &Stream{
+		transport: opts.Transport,
+		backoff:   opts.Backoff,
+		nodeProto: opts.NodeProto,
+		lrsStore:  load.NewStore(),
+	}
+
+	l := grpclog.Component("xds")
+	lrs.logger = igrpclog.NewPrefixLogger(l, opts.LogPrefix+fmt.Sprintf("[lrs-stream %p] ", lrs))
+	return lrs
+}
+
+// ReportLoad returns a load.Store that can be used to report load, and a
+// cleanup function that should be called when the load reporting is no longer
+// needed.
+//
+// The first call to ReportLoad will start the LRS streaming call. Subsequent
+// calls will increment the reference count and return the same load.Store. The
+// cleanup function will stop the LRS stream when the last reference is removed.
+func (lrs *Stream) ReportLoad() (*load.Store, func()) {
+	lrs.mu.Lock()
+	defer lrs.mu.Unlock()
+
+	cleanup := grpcsync.OnceFunc(func() {
+		lrs.mu.Lock()
+		defer lrs.mu.Unlock()
+
+		if lrs.refCount == 0 {
+			lrs.logger.Errorf("Attempting to stop an LRS stream that has already been stopped.")
+			return
+		}
+		lrs.refCount--
+		if lrs.refCount != 0 {
+			return
+		}
+		lrs.cancelStream()
+		lrs.cancelStream = nil
+		lrs.logger.Infof("Stopping LRS stream")
+	})
+
+	if lrs.refCount != 0 {
+		lrs.refCount++
+		return lrs.lrsStore, cleanup
+	}
+
+	lrs.refCount++
+	ctx, cancel := context.WithCancel(context.Background())
+	lrs.cancelStream = cancel
+	lrs.doneCh = make(chan struct{})
+	go lrs.runner(ctx)
+	return lrs.lrsStore, cleanup
+}
+
+// runner is responsible for managing the lifetime of an LRS streaming call. It
+// creates the stream, sends the initial LoadStatsRequest, receives the first
+// LoadStatsResponse, and then starts a goroutine to periodically send
+// LoadStatsRequests. The runner will restart the stream if it encounters any
+// errors.
+func (lrs *Stream) runner(ctx context.Context) {
+	defer close(lrs.doneCh)
+
+	// This feature indicates that the client supports the
+	// LoadStatsResponse.send_all_clusters field in the LRS response.
+	node := proto.Clone(lrs.nodeProto).(*v3corepb.Node)
+	node.ClientFeatures = append(node.ClientFeatures, "envoy.lrs.supports_send_all_clusters")
+
+	runLoadReportStream := func() error {
+		// streamCtx is created and canceled in case we terminate the stream
+		// early for any reason, to avoid gRPC-Go leaking the RPC's monitoring
+		// goroutine.
+		streamCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		stream, err := lrs.transport.CreateStreamingCall(ctx, "/envoy.service.load_stats.v3.LoadReportingService/StreamLoadStats")
+		if err != nil {
+			lrs.logger.Warningf("Creating new LRS stream failed: %v", err)
+			return nil
+		}
+		if lrs.logger.V(2) {
+			lrs.logger.Infof("LRS stream created")
```

Comment:
Same comment as other PR: log a property of LRS Stream to id it within binary?

Reply:
Same comment as in the other PR. There is no unique property of the stream.

Reply:
Ah ok and we actually already log a unique property of the stream as it is part of prefix logger. Sounds good.